tests/test_cmd_animation.sh

```diff
@@ -60,14 +61,24 @@ pushd ${TMP_DIR}
   "${ARE_IMAGES_EQUAL}" "${INPUT_Y4M_0}" "${DECODED_FILE}" 0 && exit 1
 
   # Lossless test.
-  "${AVIFENC}" -s 8 "${INPUT_Y4M_0}" "${INPUT_Y4M_1}" -q 100 -o "${ENCODED_FILE}"
+  "${AVIFENC}" -s 8 -q 100 "${INPUT_Y4M_0}" "${INPUT_Y4M_1}" -o "${ENCODED_FILE}"
```

Comment:
> The first occurrence of each frame flag type also applies to all frames appearing before it in the command line

This would be too conterintuitive for new users not knowing the history of `avifenc`.
We may specially apply trailing flags if there's only one input file, but also print a warning to notify the change to the flag semantic. For multiple inputs case, there seems to have no way to be both intuitive and backward compatible.

Luckily we are still at `v0.x`. @wantehchang  I see you are planning `v1.0` release in #1509. This change may need to be included in `v1.0` as well if we want it, or there probably won't be chance to do it later.

Reply:
After discussion with @wantehchang we decided to keep backward compatibility as much as possible, and thus this pull request can be merged after the libavif 1.0.0 release.

Instead of positional arguments, @wantehchang suggested frame-specific prefixes similar to the channel-specific prefixes used for the `-a` flag:

https://github.com/AOMediaCodec/libavif/blob/5e7a1778f1199b4cb96f25f64a1dc050d7e15ee9/apps/avifenc.c#L161-L162

Would something like `-q 3:75` work? Meaning `-q 75` but applied to the frame with index 3 only.

Reply:
Thank you for the detailed analysis.

> I'd still place them right before target input so I don't need to count to understand what it does - but I still need to count when writing commands.

`-q next:75` syntax is fine with me, next to `-q 3:75` syntax (for y4m input).

> The parsing have to be two-pass.

I would say that it matters less than backward compatibility and user friendliness, unless it makes the code overly complex.

> Is "frame with index 3" "output frame index 3" or "input file index 3"? I don't have clear preference, but in case we end up needing both, it will be even more complicated than the --from-index approach I wrote before.

I would go with "output frame index 3".

> I'd prefer "3" means starting from "3" if we going through this way, so it maps precisely to a programmatic change to one of `avifEncoder` field.

`-q 3+:75` syntax is fine with me (or something similar to signal that frames >=3 are targeted). It can coexist with `-q 3:75` syntax if the latter is useful.

> However, now for the most common single input case, trailing options still have effect; we can also treat the trailing options as applying to the first input for multiple input case. I would say it's a really strange expectation, if someone is putting options in between inputs and expect it to apply to all inputs - even if avifenc actually works in this way.

What about this precedence order, for each frame:

1. Apply the right-most frame-specific flag that targets at least this frame.
2. If none, apply the right-most global flag.
3. If none, apply default value.

Example: frame 3 will have final q75 for flags `-q 3:73 -q 3+:74 -q 2+:75 -q 76`

The most important part is keeping the current behavior for the currently possible ordered flag combinations. To prove that, it would be good to keep existing sh tests as is in this PR (lines can be added to them of course).

Reply:
To be honest I don't like the index-prefix approach. But I also understand the compatibility concern.

Have you considered making this an opt-in feature? We can require an environment variable, say `AVIFENC_ENABLE_CHANGE_FRAME_SETTINGS`, set to `ON`, to alter the behavior of frame flags and enable changing settings support, otherwise keep the old behavior.

Reply:
> To be honest I don't like the index-prefix approach. But I also understand the compatibility concern.

I agree, positional arguments from the start would have been better. avifenc was not supposed to be a feature-complete toolbox from what I understand, but merely a demo API usage example and a simple command line binary for trivial use cases. Maybe we should have stopped adding features a while ago but it is often tempting and convenient to do so.

If the index-prefix approach is not suitable, an alternative would be to only add higher level flags with less flexibility to implement the main progressive encoding use cases. For example there is no need to mess with flag ordering if a single new flag `--progressive` picks reasonable layer qualities and scaling etc. as long as extra tuning can be achieved through the API. One can also rely on other tools once the feature is implemented there (libheif, ffmpeg).

No matter the chosen solution, sorry for the churn.

> Have you considered making this an opt-in feature? We can require an environment variable, say `AVIFENC_ENABLE_CHANGE_FRAME_SETTINGS`, set to `ON`, to alter the behavior of frame flags and enable changing settings support, otherwise keep the old behavior.

In my opinion it would only increase the complexity and this is clearly confusing.

Reply:
> For example there is no need to mess with flag ordering if a single new flag `--progressive` picks reasonable layer qualities and scaling etc. as long as extra tuning can be achieved through the API.

Sadly there's not much move since the introduction of progressive support through API (or am I missing something?). So I hope making it accessible from command line may let more people able to try it out, like front-end developers who don't know much of the low-level C stuff. But AVIF's progressive support introduced some new possibility that old formats don't support, so we don't yet have the idea about what turns out to be the best. And users are lazy to try out if there's a somewhat OK default (hope the current `--progressive` is not going to be the widely known "somewhat OK default").

I agree with the idea to just provide a few presets with a single flag, but before that we should know that they are actually good choices. I'd like the community to try and explore good choices, but that itself requires an easy to understand way to specify different configurations. My first attempt on `--progressive` was inventing a long string to specify the whole progressive config, and that's hard to understand (I myself later also have to read the help to use it). Now the index prefixes and their interaction with non-prefixed version also starts to have that smelling, so that's why I don't like it.

Or maybe Google or other companies can do some experiments to test different progressive configurations and find the good choices, so that we can directly add them. ByteDance leans toward HEIC, so there's not much I can do on this.

> In my opinion it would only increase the complexity and this is clearly confusing.

I was hoping that it can help gradually fade out the old behavior, so that we don't have to be stucked to the design mistake indefinitely. Here are other alternatives:

- Make another executable that supports changing config, like `cwebp` and `img2webp`.
- Design a "progressive script" file do describe different progressive config, and add a flag to read it, like `cjpeg`'s `-scans FILE`. We can later make one or a few choices builtin as presets, like `cjpeg`'s `-progressive`

Reply:
Just realized the part I don't like is the index that enables changing config at anywhere from anywhere. What about allowing only one prefix that means apply to inputs after it? This should keep the compatibility. I'd prefer a prefix as short as possible through, what about `-q @75`?

Reply:
> using lists, for example -q 73,74,75

That wants options have "set current" semantic, but we're sort of locked to "update" semantic:

There's the `--advanced` option, which in general we can't (or don't know how to) unset. Set for current frame means we need to unset it (revert to default) for the next frame, but we don't know what's the default so we can't do it. Therefore for `--advanced` we are restricted to do "update", and we'd better be consistent and make every option do "update" then.

I had noticed this restriction while adding the support for the library, but forgot the exact reason why we'd do "update" for the command as well, so I was talking about being consistent with using library.

> I guess this only works if the delimiter , is not already used in a frame-specific flag.

Not likely to happen, but if someone decide to use comma in their filename: `--advanced film-grain-table=file,name.tbl` (maybe with a more reasonable name).

> How important is it to keep settings flags close to their corresponding input file path? How important is it to not have to count output frames?

It's more about being able to copypaste / reusing part of the options.

Say you come up with some adaptive logic to generate progressive images:

- For all images we have a quick preview layer and a full layer. 
- If the image is large, insert one more layer in between.

With the current approach, you can have a command segment for each layer, and just join the parts of layers that you decide to have. With explicit index or comma/somewhat separated list, you can't to this and have to assemble whole command for each case separately.

Also during experiment, that will need more modifications / more scattered modifications to the command to add / remove layers (that is not the last one, which is the usual case).

> Let me know when this PR is ready for review.

Oh It's ready now. Sorry for this.

Reply:
> It's more about being able to copypaste / reusing part of the options.
> 
> Say you come up with some adaptive logic to generate progressive images:
> 
> * For all images we have a quick preview layer and a full layer.
> * If the image is large, insert one more layer in between.
> 
> With the current approach, you can have a command segment for each layer, and just join the parts of layers that you decide to have. With explicit index or comma/somewhat separated list, you can't to this and have to assemble whole command for each case separately.
> 

In this case i don't think there is a big difference between the "list" syntax and a "positional" syntax. Either way you have to assemble the arguments and combine them into one string. Whether you assemble them as "-q 1,2,3 file1 file2 file3" or "-q 1 file1 -q 2 file2 -q 3 file3" does not really matter.


> Also during experiment, that will need more modifications / more scattered modifications to the command to add / remove layers (that is not the last one, which is the usual case).
> 

For the progressive images case, since the number of inputs is limited (if i understand correctly, there can only be a maximum of 4 inputs for 4 layers), i really think the list syntax is the simplest way to go rather than introducing positional arguments and creating backwards incompatibility for the more general use-case (single layer non-progressive avif images).

Reply:
> I guess this only works if the delimiter , is not already used in a frame-specific flag.

I am not sure what you mean by a "frame-specific flag", but we already use comma as a delimiter for `--crop`, `--clap` and `--pasp` flags.

Also, the only other positional argument that i see in avifenc is the `--duration` flag which can be specified multiple times when creating animated AVIF images. The same flag can be repeated multiple times and it applies to all the following frames unless a new value is specified with another `--duration` flag.

We can consider doing the same thing with `-q` as well. But that still introduces backwards incompatibility for the wider use-case. So i am not in favor of it unless others feel strongly so.

Reply:
> "-q 1 file1 -q 2 file2 -q 3 file3"

> introducing positional arguments and creating backwards incompatibility for the more general use-case (single layer non-progressive avif images).

> But that still introduces backwards incompatibility for the wider use-case.

"positional" behavior is enabled with the `:u` (or `:update`) suffix on the flag, so there's no compatibility issue now. That means `-q:u 1 file1 -q:u 2 file2 -q:u 3 file3`. `-q 1 file1 -q 2 file2 -q 3 file3` would behave exactly the same as the current: only `-q 3` is effective. (This PR add a warning message in this case).


> In this case i don't think there is a big difference between the "list" syntax and a "positional" syntax.

Let's consider some cases:

With "positional" syntax, tweak each layer separately is easy:

```bash
LAYER1="-q:u 50 --scaling-mode:u 1/8 $INPUT_FILE"
LAYER2="-q:u 30 --scaling-mode:u 1/2 $INPUT_FILE"
LAYER3="-q:u 80 --scaling-mode:u 1 $INPUT_FILE"

if [ $INPUT_WIDTH -gt 1024 ]; do
    avifenc -j 8 $LAYER1 $LAYER2 $LAYER3 -o $OUTPUT_FILE
else
    avifenc -j 8 $LAYER1 $LAYER3 -o $OUTPUT_FILE
fi
```

With "list" syntax, you have to spell (and modify) some option values twice:

```bash
if [ $INPUT_WIDTH -gt 1024 ]; do
    avifenc -j 8 -q 50,30,80 --scaling-mode 1/8,1/2,1 $INPUT_FILE $INPUT_FILE $INPUT_FILE -o $OUTPUT_FILE
else
    avifenc -j 8 -q 50,80 --scaling-mode 1/8,1 $INPUT_FILE $INPUT_FILE -o $OUTPUT_FILE
fi
```

Otherwise you need to make 3 variables for each option, or 6 in total for this case.

And there's also boolean flag:

```bash
# Don't tile for the 1/8 size first layer
avifenc --some-global-options \
  -q:u 50 --scaling-mode:u 1/8 input.png \
  -q:u 80 --scaling-mode:u 1 --autotiling:u input.png \
  -o output.avif
```

For list syntax, that'll need `--autotiling false,true`. (or `--no-autotiling,autotiling`?)

And for `--advanced`, list syntax can be confusing:

```bash
# Looks like setting two options at once,
# but is actually setting denoise-noise-level=50 at first input, and c:sharpness=7 at second.
avifenc -a denoise-noise-level=50,sharpness=7 input.png input.png output.avif
```

Whereas in positional syntax, it's always clear:

```bash
avifenc --some-global-options \
  -a denoise-noise-level=50 input.png \
  -a:update sharpness=7 input.png \
  output.avif
```